User interface for Manuscript information

pli888 opened the pull request:
# Pull request for issue: #1112 and #1114

This is a pull request for the following functionalities:

* User interface for viewing, creating, updating and deleting Manuscripts records.
* Configuration to instantiate Nginx and PHP-FPM Docker containers enabling manuscript information to be viewed at http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=manuscript.

## How to test?

Go to the gigareview directory then start the gigareview application
```
$ cd gigareview
$ ./up.sh
```

Fetch latest EM reports from SFTP server:
```
$ docker-compose run --rm console ./yii fetch-reports/fetch
```

Process jobs in manuscript queue:
```
$ docker-compose run --rm console ./yii manuscripts-q/run
```

Go to http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=manuscript. If there were manuscripts that have been accepted for publication during the previous day then these will be displayed on the Manuscripts page.

## How have functionalities been implemented?

The `docker-compose.yml` file contains 2 new services: `web` and `application`. The `web` service provides a Nginx web server whose container is built using the latest Nginx image. The `application` service is a PHP-FPM application server built using `common/Dockerfile`.

The view classes for Manuscript and the `frontend/controllers/ManuscriptController.php` file were generated using the Gii tool.

## Any issues with implementation?

The Gii tool has been left available at http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=gii. Use of Gii required extra configuration in `frontend/config/main.php`

## Any changes to automated tests?

No extra tests were created. Acceptance tests for the functionality in this PR will be developed in #1113.

## Any changes to documentation?

The `gigareview/README.md` file has been updated with information on how to download and view Manuscript information.

pli888 committed:
Enable use of Gii tool

Gii tool available at http://localhost/frontend/web/index.php?r=gii

pli888 committed:
Add bootstrap dependency

pli888 committed:
Add manuscript view classes

pli888 committed:
Enable access at gigareview.gigasciencejournal.com:9180

pli888 committed:
Re-create classes with I18N and Pjax enabled

pli888 committed:
Add documentation on how to use GigaReview

pli888 committed:
Add application container

The application container provides a PHP FPM application server
for GigaReview.

pli888 committed:
Add configuration code to create AppAsset.php

pli888 committed:
Make GigaReview dev accessible from gigadb.gigasciencejournal.com:9170

pli888 committed:
Update manuscript view classes

Manuscript view classes updated due to new editors_note column.

kencho51 commented on gigareview/frontend/views/manuscript/index.php:
```diff
@@ -0,0 +1,51 @@
+            'manuscript_number',
+            'article_title',
+            'publication_date',
+            //'editorial_status',
```
@pli888 Is it better uncomment these to  make them visible in manuscript index page?

kencho51 replied:
@pli888 Is it better uncomment these to make them visible in manuscript index page?

kencho51 commented on gigareview/frontend/views/manuscript/index.php:
```diff
@@ -0,0 +1,51 @@
+            'manuscript_number',
+            'article_title',
+            'publication_date',
+            //'editorial_status',
+            //'editorial_status_date',
+            //'editors_note:ntext',
```
@pli888 Is it better uncomment these to make them visible in manuscript index page?

pli888 replied:
These 3 fields have been uncommented now.

kencho51 commented on gigareview/frontend/views/manuscript/_form.php:
```diff
@@ -0,0 +1,39 @@
+
+    <?= $form->field($model, 'editorial_status')->textInput(['maxlength' => true]) ?>
+
+    <?= $form->field($model, 'editorial_status_date')->textInput() ?>
+
```
@pli888 The form could not be updated unless the `Editorial Status Date` changes to `07/06/2022`, I think the reason for this is because in the date format of this column is restricted to `dd/MM/yyyy` in `gigareview/common/models/Manuscript.php`.

pli888 replied:
The rule for `editorial_status_date` (as well as for `publication_date`) has been updated so that the dates from EM Manuscript reports are in the form of MM/dd/yyyy.

rija commented on gigareview/docker-compose.yml:
```diff
@@ -2,10 +2,34 @@ version: '3.7'
+      - ./:/app
+      - ./default.conf:/etc/nginx/conf.d/default.conf
+
+  application:
```
Hi @pli888,
container services ``application`` should not exit.
Since this is an Yii2 advanced template, there are two application servers: ``frontend`` and ``backend`` (respectively called ``public`` and  ``api`` in our config here). 
The ``public`` is the one to be used for public access by web user, so that's the one that should serve the manuscript dashboard.

rija commented on gigareview/docker-compose.yml:
```diff
@@ -2,10 +2,34 @@ version: '3.7'
+    build:
+      context: .
+      dockerfile: ./common/Dockerfile
+    ports:
```
On application servers, this block is unnecessary

```
    ports:
      - "9009:9000"
```

the reason:

* all services inside a docker-compose have access to the same network and their internal exposed ports is accessible to eachother 
* There no reason to access the application server from outside Docker (that's what `9009` allows), as only the web container service needs to acces it

rija commented on gigareview/docker-compose.yml:
```diff
@@ -2,10 +2,34 @@ version: '3.7'
 services:
 
+  web:
+    image: nginx:latest
```
It's better to use a specific version number instead of ``latest``, 
because the value of  ``latest`` changes with each new releases, which creates two problems:

* a new version may introduce bugs
* our infrastructure is no longer immutable, as infrastructure elements may change independently of CI/CD pipeline deployment

rija commented:
Hi @pli888, @kencho51 

When navigating to ``http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=manuscript`` as instructed in the "How to test" section, I have a 500 server error returned. I'm looking at possible cause, but unsuccessful so far:
![Screenshot 2022-10-04 at 11 59 08 AM](https://user-images.githubusercontent.com/11455/193800078-9659fcb6-72fb-4de2-bcd7-a69b2c5dc6e8.png)

kencho51 commented:
Hi @rija,

> When navigating to `http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=manuscript` as instructed in the "How to test" section, I have a 500 server error returned. I'm looking at possible cause, but unsuccessful so far

This error does not happen in my case, and I could get gigareview up and running:
<img width="1135" alt="image" src="https://user-images.githubusercontent.com/64770635/193857692-1abacb47-dba3-4a33-8b85-bab972899908.png">

rija commented:
Hi @pli888, @kencho51 

I've deleted my local clone, and local database, and started from scratch, and I still get the same server error.
I will continue investigating.

Also there is an error thrown up by  up.sh if ``gigareview/.env`` doesn't exist,  remediated by adding:

```
CI_PROJECT_URL="https://gitlab.com/api/v4/projects/gigascience/forks/$REPO_NAME/"
```
in ``config-sources/env.example``, after the line containing ``REPO_NAME=``

pli888 commented:
>When navigating to http://gigadb.gigasciencejournal.com:9170/gigareview/index.php?r=manuscript as instructed in the "How to test" section, I have a 500 server error returned. I'm looking at possible cause, but unsuccessful so far:

@rija A couple of error messages after executing `./up.sh` shows what the problem is:
```
  Error. Directory backend/web/assets does not exist. 
      chmod 0777 console/runtime
      chmod 0777 frontend/runtime

  Error. Directory frontend/web/assets does not exist. 
      chmod 0755 yii
      chmod 0755 yii_test

  ... initialization completed.
```

Creating the 2 `assets` directories fixes the problem. I'll include a fix into the PR.

pli888 committed:
Add CI_PROJECT_URL variable

pli888 committed:
Create assets directories

These 2 assets directories are ignored by git.

pli888 committed:
Use nginx version 1.21.3 in web container service

Version 1.21.3 NGINX is what is used for GigaDB.

pli888 committed:
Remove ports blocks

No need to expose these ports for application containers outside of
Docker.

pli888 committed:
Remove application container service

We use the public container as the application server for public
access by the web user. It is this application server that will
serve the manuscript dashboard.

pli888 committed:
Display more manuscript fields

pli888 committed:
Update rule for editorial_status_date

The editorial_status_date in EM manuscript reports is in the form
of month/day/year.

pli888 committed:
Expose port 9000 from api container

Only need to expose port 9000 from api container. No need to make
it available from port 9011 on localhost.

pli888 committed:
Stop access to web-based Gii tool

pli888 committed:
Create unit test to check format of editorial status date

pli888 committed:
Update testing of editorial status dates

pli888 committed:
Validate manuscripts before saving

If we don't validate before saving then invalid attribute data
could be stored in database and result in future errors.

pli888 committed:
Improve invalid date format test

A second mock manuscript containing editorial_status_date in
invalid format has been added. It cannot be saved into database
due to so test will pass.

pli888 commented:
I've now done the list of things below:

- [x] Display editorial_status, editorial_status_date, editors_note fields on manuscript index page
- [x] Update date format validation rule from `dd/MM/yyyy` to `MM/dd/yyyy` in `Manuscript` model class to correspond to date in EM manuscript reports
- [x] Use `public` container as the application server
- [x] `public` and `api` application server ports are not exposed ports outside of Docker
- [x] Use specific 1.21.3 version for nginx image in `docker-compose.yml` - this version is  same as main GigaDB application
- [x] Create directories `backend/web/assets` and `frontend/web/assets` in `generate_config.sh`
- [x] Add `CI_PROJECT_URL` variable into `config-sources/env.example`
- [x] Remove configuration for web-based Gii tool
- [x] Update `testEditorialStatusDateValidation` unit test to check date validation in Manuscript objects

In addition, manuscript instances are validated before saving in `storeManuscripts()` in `EMReportJob`. Without validation, dates such as 13/06/2022 could be stored in database and lead to produce bugs since dates are expected to be in MM/dd/yyyy format. If manuscript instances fail validation then they are not saved and will output an error message:
```
$ docker-compose run --rm console ./yii manuscripts-q/run
Creating review_console_run ... done
Error: Manuscript GIGA-D-22-00078 failed validation
array(1) {
  ["editorial_status_date"]=>
  array(1) {
    [0]=>
    string(47) "The format of Editorial Status Date is invalid."
  }
}
```

If the validate or save function ever fails, I would think we would need to be notified about this?

The unit test `testCannotStoreInstanceWithInvalidDateFormatToManuscriptTable()` in `EMReportJobTest` has been improved to check that manuscripts with editorial_status_date with dd/MM/yyyy format cannot be saved into database due to a prior validation step.

rija commented:
@pli888
praise: quick note to confirm that your latest change has fixed the blocking problem I had.

Thanks for that. I'll pick up full review again next week when I'm back home and my hiring tasks are behind.

---
https://conventionalcomments.org/

kencho51 commented on gigareview/common/models/EMReportJob.php:
```diff
@@ -87,8 +87,15 @@ public function storeManuscripts(array $manuscripts): bool
-            if ($manuscript->save()) {
-                $storeStatus = 1;
+            # Check all manuscript attribute values are valid before saving
+            if ($manuscript->validate()) {
```
Hi @pli888,

Since the manuscript object will be validated at this step, if the sftp server config in `gigareview/common/config/params-local.php` changed to
```
    'sftp' => [
        "host" => "sftp_test",
        "username" => "testuser",
        "password" => "testpass",
        "baseDirectory" => "editorialmanager",
    ],
```
and when executing the manuscript-q job, would get the validation error:
```
 % docker-compose run --rm console ./yii manuscripts-q/run`
Creating review_console_run ... done
Error: Manuscript GIGA-D-22-00060 failed validation
array(1) {
  ["editorial_status"]=>
  array(1) {
    [0]=>
    string(28) "Editorial Status is invalid."
  }
}
Error: Manuscript GIGA-D-22-abcde failed validation
array(1) {
  ["manuscript_number"]=>
  array(1) {
    [0]=>
    string(29) "Manuscript Number is invalid."
  }
}
Error: Manuscript GIGA-D-22-00099 failed validation
array(1) {
  ["editorial_status_date"]=>
  array(1) {
    [0]=>
    string(47) "The format of Editorial Status Date is invalid."
  }
}
```
So, maybe is it better to remove those invalid data in `gigareview/console/tests/_data/Report-GIGA-em-manuscripts-latest-214-20220607004243.csv` which were meant to test the validation rules in `manuscript` model class.

pli888 replied:
@kencho51 I've corrected the validation errors in the CSV file. I've written new unit tests to check manuscript number and editorial status errors. There is also new documentation on how to use the `sftp_test` service in `README.md`.

After looking at my code again, I realise acceptance tests for manuscript records on the manuscript index page are missing from this PR.

pli888 committed:
Provide documentation to use sftp_test SFTP docker service

pli888 committed:
Add new unit tests

These tests check editorial statuses and manuscript numbers can be
correctly validated.

pli888 committed:
Update manuscript record data

Fixed validation errors in manuscript records.
